plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -110,12 +160,23 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
          derived_session().socket_stream->binary(true);
          derived_session().socket_stream->next_layer().set_option(boost::asio::socket_base::send_buffer_size(1024 * 1024));
          derived_session().socket_stream->next_layer().set_option(boost::asio::socket_base::receive_buffer_size(1024 * 1024));
-         derived_session().socket_stream->async_accept([self = derived_session().shared_from_this()](boost::system::error_code ec) {
-            self->callback(ec, "async_accept", [self] {
-               self->start_read();
-               self->send(state_history_plugin_abi);
+         derived_session().socket_stream->async_accept(
+            [self = derived_session().shared_from_this()](boost::system::error_code ec) {
+               self->callback(ec, "async_accept", [self] {
+                  self->socket_stream->binary(false);
+                  self->socket_stream->async_write(
+                     boost::asio::buffer(state_history_plugin_abi, strlen(state_history_plugin_abi)),
+                     [self](boost::system::error_code ec, size_t) {
+                        if (ec) {
+                           self->on_fail(ec, "async_write");
+                           return;
+                        }
+                        self->socket_stream->binary(true);
+                        self->start_read();
```

Comment:
I'm pretty sure this is not thread safe. You are not allowed to access a socket from more than one thread at a time. Here you have the main thread doing an `async_read` and the internal thread doing an `write`.

Reply:
That's not what I understand about socket programming. According to https://stackoverflow.com/questions/7418093/use-one-socket-in-two-threads, it should be safe for TCP. AFAIK, websocket is full-duplex communication channels over a single TCP connection, it should also be safe for read/write from different threads.

Reply:
See thread-safety at the bottom of the page: https://www.boost.org/doc/libs/1_76_0/doc/html/boost_asio/reference/ip__tcp/socket.html
Also: 
https://stackoverflow.com/questions/7362894/boostasiosocket-thread-safety

Reply:
If you really want to do it, you can dup() the fd and create two boost sockets referring to the same underlying TCP connection. Then use one boost socket for read, one for write, on two separate threads. Not that i'm recommending doing that here, but that's how it can be done.

Reply:
In case anyone finds this: my brain was thinking asio socket. The semantics of dup()ing a socket backed by a websocket::stream are almost certainly no good!